### ISSUE TITLE
InputText() for capturing non-printable chars/keys

### ISSUE THREAD

**Issue by jdryg** [ocornut/imgui#1288]
Hi,

This is more of a question on where to start looking rather than a request for an implementation.

What I need for my app is a control similar to InputText() which will be able to capture non-printable characters, such as Backspace, Delete, arrow keys, Home/End, Tab and Enter. 

I.e. when Backspace is pressed, instead of deleting the character before the cursor, 0x08 will be inserted into the buffer (0x08 == BACKSPACE ASCII value). The same will be true for all the other captured keys. If there's no ASCII value for one of those, I could use an arbitrary/unused ASCII value.

I don't currently care about correctly rendering those characters, though I suspect that a custom font with correct glyphs at those ASCII values will work. Am I right?

Is this possible? Can it be implemented as a widget or it'll be better to somehow implement it outside of imgui (i.e. a read-only InputText() which when focused, the app will manually manipulate the buffer)?

Thanks :)

**Comment by jdryg** [comment]
I trimmed down InputTextEx() a bit to help me figure out what's going on. Ended up replacing the special cases for all the keys I want and it seems to work.

Nothing really interesting but here's my code: https://gist.github.com/jdryg/a4dc8d9f75ae298f0754c7ee39c457cd

Now to an imgui-unrelated issue on how to convert those chars to console commands :)

**Comment by ocornut** [comment]
You don't really need most of that code you copied.. In fact if you think about it, your problem has _little_ overlap with InputText at all and you shouldn't be using an InputText nor that code as a base.

What you want is:
- Read keyboard input and store that in a format you like (your Backspace key -> 0x08)
- Display value in those format in a format you like (0x08 -> "Backspace").

**Comment by ocornut** [comment]
You don't really need most of that code you copied.. In fact if you think about it, your problem has _little_ overlap with InputText at all and you shouldn't be using an InputText nor that code as a base.

Do you really need text edition capatibilities here? What are you using it for? Input mapping? If so, what you want may be more:

- Read keyboard input and store that in a format you like (your Backspace key -> 0x08)
- Display value in those format in a format you like (0x08 -> "Backspace").

**Comment by jdryg** [comment]
Ideally what i want is a control which *looks* like InputText() but behaves like the code I posted. 
I.e. always append chars to the buffer and don't translate them to cursor operations.

Of course I don't need (e.g.) selection, but I thought it'd be easier to just start with what's already there :)

My initial approach was to use a read-only InputTextEx() and if it had keyboard focus (IsItemActive()) handle the input buffer manually. Problem is that InputTextEx() consumes all the chars and the app doesn't receive and input. So it had to be a separate control.

Hope that makes sense. I'll try to reduce it further at some point but it works for now, so there's no need for something else.

**Comment by ocornut** [comment]
> I thought it'd be easier to just start with what's already there :)

`InputText()` is unfortunately quite a minefield at the moment, I wouldn't recommend copying so much of its code.

I still don't quite understand your use case. Keyboard input would be consumed immediately by the emulator. It is more of a log of keys pressed, in which case you can just log into a text buffer and display that with `TextWrapped()` ? Or it is more like a macro editor/recorder?

**Comment by jdryg** [comment]
I guess a picture's worth a 1000 words :)

![riscv_emulator_stdinout](https://user-images.githubusercontent.com/5159208/29485354-3c7bcc16-84d9-11e7-928d-b85532f68e32.png)

TL;DR:
The reason I wanted keyboard input to pass through imgui is because I want it to be transmitted to the emulator only when stdin is focused. Since imgui already gets all the required info from GLFW's callbacks I thought it'd be better to do it this way (instead of e.g. adding special cases in GLFW's callbacks in order to capture keyboard input in another buffer).

The 1000 word description:
When the stdin control in the image is focused, all key presses should be recorded in a buffer. Every frame, the emulator sends the first char of the buffer (if there's any) to the emulated CPU via a simple UART. 

The code running on the emulated CPU polls the UART for input. If it doesn't poll, no char is transmitted on the next frame (UART has only 1 data register and no FIFO buffer). 

If the emulated code reads the UART, the emulator is free to send the next char to it. It's the emulated code's responsibility to translate the char to whatever it wants.

For a simple gets()/readline() implementation reading ASCII chars from the UART, filling a buffer and echoing the char is enough. So an InputTextEx() works. Until you have to handle backspace and other special chars. InputTextEx() consumes those chars and translates them to cursor operations, and the emulator doesn't get any notification about those key presses.

Does it make a bit more sense now? :)

DISCLAIMER: I might be wrong about how all of these stuff should work (keyboards, polling, etc) in real circuits. I'm still learning as I go.

**Comment by ocornut** [comment]
So it looks more like a FIFO.
Why not just recording the input in a buffer (either in text format or in raw format), and have a function that render that buffer as text? You barely need any of the `InputText` code there.

**Comment by jdryg** [comment]
Is there another build-in control in imgui which is focusable and displays text, except InputText()? 

I know I should be able to implement that, but as I said it seemed easier to start with an existing control (I haven't really tried to understand imgui's code and I don't know what's required to implement even the simplest of controls, that's why butchering existing code sounds easier for me :))

**Comment by jdryg** [comment]
Removed everything related to edit_state and selections. Just a simple frame with some text on it and a blinking cursor (always at the end of the string).

https://gist.github.com/jdryg/a4dc8d9f75ae298f0754c7ee39c457cd
